Cubical/ZCohomology/RingStructure/RingLaws.agda

```diff
@@ -0,0 +1,620 @@
+      fst (⌣ₖ-distrFun-r (suc n) (suc m) ∣ x ∣ ∣ ptSn (suc n) ∣) ≡
+      fst (⌣ₖ-distrFun2-r (suc n) (suc m) ∣ x ∣ ∣ ptSn (suc n) ∣)
+    fst-right n x =
+      funExt λ z
+        → cong (_⌣ₖ z) (rUnitₖ _ ∣ x ∣)
+        ∙∙ sym (rUnitₖ _ _)
+        ∙∙ cong (∣ x ∣ ⌣ₖ z +ₖ_) (sym (0ₖ-⌣ₖ _ _ z))
+
+    left≡right : (n : ℕ) → fst-left n (ptSn (suc n)) ≡ fst-right n (ptSn (suc n))
+    left≡right zero = refl
+    left≡right (suc n) = refl
+
+    main : (a b : S₊ (suc n)) →
+      ⌣ₖ-distrFun-r (suc n) (suc m) ∣ a ∣ₕ ∣ b ∣ₕ ≡
+      ⌣ₖ-distrFun2-r (suc n) (suc m) ∣ a ∣ₕ ∣ b ∣ₕ
+    main =
+      wedgeconFun n n
+        (λ x y → subst (λ l → isOfHLevel l ((⌣ₖ-distrFun-r (suc n) (suc m) ∣ x ∣ ∣ y ∣)
+                                           ≡  ⌣ₖ-distrFun2-r (suc n) (suc m) ∣ x ∣ ∣ y ∣))
+                        (+-suc n (suc n))
+                        (isOfHLevelPlus {n = 2 + n} n (isOfHLevel↑∙ (suc n) m _ _)))
+        (λ x → →∙Homogeneous≡ (isHomogeneousKn _) (fst-left n x))
+        (λ x → →∙Homogeneous≡ (isHomogeneousKn _) (fst-right n x))
+        (cong (→∙Homogeneous≡ (isHomogeneousKn _)) (sym (left≡right n)))
+
+rightDistr-⌣ₖ : (n m : ℕ) (x y : coHomK n) (z : coHomK m)
+          → (x +ₖ y) ⌣ₖ z ≡ (x ⌣ₖ z +ₖ y ⌣ₖ z)
+rightDistr-⌣ₖ zero zero x y z =
+     comm-·₀ (x ℤ+ y) z
+  ∙∙ leftDistr-⌣ₖ zero zero z x y
+  ∙∙ cong₂ _+ₖ_ (sym (comm-·₀ x z)) (sym (comm-·₀ y z))
+rightDistr-⌣ₖ zero (suc m) x y z = leftDistr-⌣ₖ _ zero z x y
+rightDistr-⌣ₖ (suc n) zero x y z = leftDistr-⌣ₖ zero (suc n) z x y
+rightDistr-⌣ₖ (suc n) (suc m) x y z = (funExt⁻ (cong fst (rightDistr-⌣ₖ· n m x y))) z
+
+-- Associativity
+
+private
+-- We need to give the two associators as (doubly) pointed functions
+  assocer : (n m k : ℕ) → coHomK (suc n)
+    →  coHomK-ptd (suc m)
+    →∙ (coHomK-ptd (suc k)
+    →∙ coHomK-ptd ((suc n) +' ((suc m) +' (suc k))) ∙)
+  fst (fst (assocer n m k x) y) z = x ⌣ₖ (y ⌣ₖ z)
+  snd (fst (assocer n m k x) y) = cong (x ⌣ₖ_) (⌣ₖ-0ₖ _ (suc k) y) ∙ ⌣ₖ-0ₖ _ _ x
+  snd (assocer n m k x) =
+    ΣPathP (funExt (λ z → cong (x ⌣ₖ_) (0ₖ-⌣ₖ (suc m) (suc k) z) ∙ ⌣ₖ-0ₖ (suc n) _ x)
+          , help)
+    where
+    h : (n m k : ℕ) (x : coHomK (suc n)) → cong (_⌣ₖ_ x) (⌣ₖ-0ₖ (suc m) (suc k) (0ₖ _)) ≡
+                                  (λ i → x ⌣ₖ 0ₖ-⌣ₖ (suc m) (suc k) (0ₖ _) i)
+    h zero zero k x = refl
+    h zero (suc m) k x = refl
+    h (suc n) zero k x = refl
+    h (suc n) (suc m) k x = refl
+
+    help : PathP (λ i → (cong (x ⌣ₖ_) (0ₖ-⌣ₖ (suc m) (suc k) (0ₖ _)) ∙ ⌣ₖ-0ₖ (suc n) _ x) i ≡ 0ₖ _)
+               (cong (x ⌣ₖ_) (⌣ₖ-0ₖ (suc m) (suc k) (0ₖ _)) ∙ ⌣ₖ-0ₖ (suc n) _ x) refl
+    help = compPathR→PathP
+      (cong (_∙ ⌣ₖ-0ₖ (suc n) ((suc m) +' (suc k)) x) (h n m k x)
+      ∙∙ rUnit _
+      ∙∙ cong ((cong (x ⌣ₖ_) (0ₖ-⌣ₖ (suc m) (suc k) (0ₖ _)) ∙ ⌣ₖ-0ₖ (suc n) _ x) ∙_) (rUnit refl))
+
+  assoc2-sub : (n m k : ℕ) → _ → _
+  assoc2-sub n m k = subst coHomK (sym (+'-assoc n m k))
+
+  assoc2-sub-0 : (n m k : ℕ) → assoc2-sub n m k (0ₖ _) ≡ 0ₖ _
```

Comment:
Yikes, can't one write this kind of thing nicer? Maybe @ecavallo or @Saizan has some ideas?